ap/columnar_util.py

```diff
@@ -842,14 +862,53 @@ def all_shifts(self):
 
```

Comment:
Shoudn't the line above be `shifts |= obj.all_shifts` to properly resolve all dependent shifts?

Reply:
Yes it should. And thinking about it, one shouldn't loop here over the `uses` and `produces` sets, but just look for other `ArrayFunction`s in `self.shifts`. This requires adding a dependent function in (e.g.) both `produces` _and_ `shifts`, but at the same time provides a better level of configurability.

Reply:
Thinking about this further, we should leave these special cases for the future and only take action if required.

Reply:
Hm, I think it actually makes sense to also loop over `produces` and `uses` here. To my understanding, one would always want the `shifts` from these dependencies to be propagated to the current function's `all_shifts`, since they indirectly affect the result of the current function as well.

I would add them for now, but we should keep this in mind in case we need to make this more flexible for some future use case.